library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java

```diff
@@ -398,37 +420,23 @@ public void put(Slice key, long timeBucket, Slice value)
       }
     }
 
+    /**
+     * free the memory which window id up to input windowId
+     */
     @Override
     public long freeMemory(long windowId) throws IOException
     {
-      long memoryFreed = 0;
-      Long clearWindowId;
-
-      while ((clearWindowId = committedData.floorKey(windowId)) != null) {
-        Map<Slice, BucketedValue> windowData = committedData.remove(clearWindowId);
-
-        for (Map.Entry<Slice, BucketedValue> entry: windowData.entrySet()) {
-          memoryFreed += entry.getKey().length + entry.getValue().getValue().length;
-        }
-      }
-
-      fileCache.clear();
-      if (cachedBucketMetas != null) {
-
-        for (BucketsFileSystem.TimeBucketMeta tbm : cachedBucketMetas) {
-          FileAccess.FileReader reader = readers.remove(tbm.getTimeBucketId());
-          if (reader != null) {
-            memoryFreed += tbm.getSizeInBytes();
-            reader.close();
-          }
-        }
-
-      }
-      sizeInBytes.getAndAdd(-memoryFreed);
-      LOG.debug("space freed {} {}", bucketId, memoryFreed);
+      long originSize = keyStream.size() + valueStream.size();
+      keyStream.resetUpToWindow(windowId);
```

Comment:
This gets called from a different thread, won't this cause a synchronization issue in WindowableBlockStream?